### PR TITLE
New routes&groups syntax

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

First syntax
```php
Group::create('/api')
        ->middleware(FormatDataResponseAsXml::class)
        ->middleware(ApiDataWrapper::class)
        ->routes(
            Route::get('/info/v1')
                ->name('api/info/v1')
                ->action(function (DataResponseFactoryInterface $responseFactory) {
                    return $responseFactory->createResponse(['version' => '1.0', 'author' => 'yiisoft']);
                }),
            Route::get('/info/v2')
                ->name('api/info/v2')
                ->middleware(FormatDataResponseAsJson::class)
                ->action(ApiInfo::class),
            Route::get('/user')
                ->name('api/user/index')
                ->action([ApiUserController::class, 'index']),
            Route::get('/user/{login}')
                ->name('api/user/profile')
                ->middleware(FormatDataResponseAsJson::class)
                ->action([ApiUserController::class, 'profile'])
        ),
```

Second syntax
```php
 Group::create('/api')
        ->routes(
            Route::get('/info/v1')
                ->name('api/info/v1')
                ->action(function (DataResponseFactoryInterface $responseFactory) {
                    return $responseFactory->createResponse(['version' => '1.0', 'author' => 'yiisoft']);
                }),
            Route::get('/info/v2')
                ->name('api/info/v2')
                ->action(ApiInfo::class)
                ->addMiddleware(FormatDataResponseAsJson::class),
            Route::get('/user')
                ->name('api/user/index')
                ->action([ApiUserController::class, 'index']),
            Route::get('/user/{login}')
                ->name('api/user/profile')
                ->action([ApiUserController::class, 'profile'])
                ->addMiddleware(FormatDataResponseAsJson::class)
        )
        ->addMiddleware(ApiDataWrapper::class)
        ->addMiddleware(FormatDataResponseAsXml::class),
```

Related PRs:

- https://github.com/yiisoft/yii-debug-api/pull/26
- https://github.com/yiisoft/yii-demo/pull/269
- https://github.com/yiisoft/app/pull/145
- https://github.com/yiisoft/app-api/pull/59
- https://github.com/yiisoft/demo-api/pull/22

### CODE REVIEW

File: src/Route.php [hunk 1]
```diff
@@ -242,6 +242,20 @@ public function addMiddleware($middlewareDefinition): self
         return $route;
     }
 
+    public function pipe($middlewareDefinition): self
```
Comment (samdark) [inline src/Route.php:245]:
```suggestion
    public function middleware($middlewareDefinition): self
```

File: src/Group.php [hunk 2]
```diff
@@ -118,6 +146,12 @@ public function addMiddleware($middlewareDefinition): self
         return $this;
     }
 
+    public function pipe($middlewareDefinition): self
```
Comment (samdark) [inline src/Group.php:149]:
```suggestion
    public function middleware($middlewareDefinition): self
```

File: src/Group.php [hunk 3]
```diff
@@ -67,6 +67,34 @@ public static function create(?string $prefix = null, $routes = [], MiddlewareDi
         return new self($prefix, $callback, $dispatcher);
     }
 
+    public function routes($routes): self
```
Comment (samdark) [inline src/Group.php:70]:
```suggestion
    public function routes(...$routes): self
```

File: src/Group.php [hunk 4]
```diff
@@ -67,6 +67,34 @@ public static function create(?string $prefix = null, $routes = [], MiddlewareDi
...
+        } elseif (is_array($routes)) {
+            $callback = static function (self $group) use (&$routes) {
+                foreach ($routes as $route) {
+                    if ($route instanceof Route) {
+                        $group->addRoute($route);
+                    } elseif ($route instanceof self) {
+                        $group->addGroup($route);
+                    } else {
+                        $type = is_object($route) ? get_class($route) : gettype($route);
+                        throw new InvalidArgumentException(sprintf('Route should be either instance of Route or Group, %s given.', $type));
```
Comment (samdark) [inline src/Group.php:83]:
```suggestion
                        throw new InvalidArgumentException(sprintf('Route should be either an instance of Route or Group, %s given.', $type));
```

**Review by samdark — APPROVED** [review]

**Comment by rustamwin** [comment]
What about doing `Route::$action` instead of `Route::$actionAdded`?

**Comment by samdark** [comment]
@rustamwin that's not about action itself, it's a boolean flag for event happened.

**Comment by samdark** [comment]
👍

**Pull request merged**